src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -492,6 +492,355 @@ TEST(Aktualizr, FullWithUpdatesNeedReboot) {
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+
+    auto aktualizr_cycle_thread = aktualizr.RunForever();
+    auto aktualizr_cycle_thread_status = aktualizr_cycle_thread.wait_for(std::chrono::seconds(10));
+
+    ASSERT_EQ(aktualizr_cycle_thread_status, std::future_status::ready);
+    ASSERT_TRUE(aktualizr.uptane_client_->bootloader->rebootDetected());
+    ASSERT_TRUE(event_hdlr.checkRecievedEvents(expected_event_order));
+    ASSERT_TRUE(aktualizr.uptane_client_->hasPendingUpdates());
+    ASSERT_TRUE(http_server_mock->checkRecievedReports(expected_report_order));
+    // Aktualizr reports to a server that installation was successfull in spite the fact that
+    // the installation requires `finalization`
+    ASSERT_TRUE(http_server_mock->wasInstallSuccessfull());
+
+    data::InstallationResult dev_installation_res;
+    std::string report;
+    std::string correlation_id;
+    ASSERT_TRUE(storage->loadDeviceInstallationResult(&dev_installation_res, &report, &correlation_id));
+    // Used to return kInstallFailed while it should be kNeedCompletion
+    EXPECT_EQ(dev_installation_res.result_code.num_code, data::ResultCode::Numeric::kNeedCompletion);
+
+    std::vector<std::pair<Uptane::EcuSerial, data::InstallationResult>> ecu_installation_res;
+    EXPECT_TRUE(storage->loadEcuInstallationResults(&ecu_installation_res));
+    EXPECT_EQ(ecu_installation_res.size(), 2);
+
+    for (const auto& ecu_install_res : ecu_installation_res) {
+      auto ecu_id = ecu_install_res.first.ToString();
+
+      if (ecu_id == primary_ecu_id) {
+        EXPECT_EQ(ecu_install_res.second.result_code.num_code, data::ResultCode::Numeric::kNeedCompletion);
+        EXPECT_EQ(ecu_install_res.second.success, false);
+        EXPECT_EQ(ecu_install_res.second.description, "Application successful, need reboot");
+
+      } else if (ecu_id == secondary_ecu_id) {
+        EXPECT_EQ(ecu_install_res.second.result_code.num_code, data::ResultCode::Numeric::kOk);
+        EXPECT_EQ(ecu_install_res.second.success, true);
+        EXPECT_EQ(ecu_install_res.second.description, "");
+
+      } else {
+        FAIL() << "Unexpected ECU serial: " << ecu_install_res.first;
+      }
+    }
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+    ASSERT_EQ(installed_versions.size(), 1);
+    // why valid
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    // why 0 if should be 1/true
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(secondary_ecu_id, &installed_versions, &current_version, &pending_version));
+    ASSERT_EQ(installed_versions.size(), 1);
+    ASSERT_TRUE(installed_versions[0].IsValid());
+    ASSERT_EQ(pending_version, 0);
+    // why 0, this is secondary that was succesfully installed, so it should 1/true ???
+    ASSERT_EQ(current_version, 0);
```

Comment:
Probably a bug, although perhaps this doesn't get set until the entire install operation (for all ECUs) has succeeded?

Reply:
See the above comment.